src/sqlancer/ASTBasedReducer.java

```diff
@@ -0,0 +1,470 @@
+            });
+
+            expressionReduce(whenClause, then, (wc, t) -> {
+                wc.setThenExpression(t);
+                updateStatements();
+            });
+
+            super.visit(whenClause);
+        }
+
+        @Override
+        public void visit(Parenthesis parenthesis) {
+            Expression closedExpr = parenthesis.getExpression();
+            expressionReduce(parenthesis, closedExpr, (p, s) -> {
+                p.setExpression(s);
+                updateStatements();
+            });
+            closedExpr.accept(this);
+        }
+
+        @Override
+        public void visit(Function function) {
+            super.visit(function);
+        }
+
+        @Override
+        public void visit(ExpressionList expressionList) {
+            List<Expression> expressions = expressionList.getExpressions();
+            listElementRemovingReduce(expressionList, expressions, (l, es) -> {
+                l.setExpressions(es);
+                updateStatements();
+            });
+            expressions = expressionList.getExpressions();
+            for (int i = 0; i < expressions.size(); i++) {
+                Expression expr = expressions.get(i);
+                int index = i;
+                expressionReduce(expressions, expr, (l, e) -> {
+                    l.set(index, e);
+                    updateStatements();
+                });
+            }
+            super.visit(expressionList);
+        }
+
+        @Override
+        public void visit(SelectExpressionItem selectExpressionItem) {
+            Expression expr = selectExpressionItem.getExpression();
+            expressionReduce(selectExpressionItem, expr, (item, e) -> {
+                item.setExpression(e);
+                updateStatements();
+            });
+            super.visit(selectExpressionItem);
+        }
+    };
+
+    SelectVisitorAdapter selectReducerVisitor = new SelectVisitorAdapter() {
+        private final String[] removeList = { "Limit", "Offset", "Where", "Having", "GroupBy", "Distinct",
+                "OrderByElements", "Joins" };
+        private final String[] pullUpList = { "Where", "Having", "FromItem", "SelectItems", "GroupBy", "Joins" };
+        private final String[] descendList = { "Where", "Having", "FromItem", "SelectItems", "GroupBy" };
+
+        private String getterName(String astNodeName) {
+            return "get" + astNodeName;
+        }
+
+        private String setterName(String astNodeName) {
+            if (astNodeName.equals("GroupBy")) {
+                return "set" + astNodeName + "Element";
+            } else {
+                return "set" + astNodeName;
+            }
+        }
+
+        @Override
+        public void visit(PlainSelect plainSelect) {
+            // Remove Section
+            for (String astNodeName : removeList) {
+                try {
+                    Method nodeGetter = plainSelect.getClass().getMethod(getterName(astNodeName));
```

Comment:
Based on my understanding, we are using reflection because we cannot change the fields otherwise, right? If so, perhaps, we could also add a comment mentioning this somewhere

Reply:
Relection is used to avoid repetitive code. For example, when trying removing clauses of an Select statement, removing `when`, `limit`, `offset` requires almost the same procedure. But the getter, setter method is not the same. I have to repeat the same code but replace the method name. e.g.: `getWhen, getLimit, getOffset`. Using reflection would make this procedure simpler.